### PR TITLE
fix: handle "receiver" in paymentSuccesNotification if used via popup #770

### PR DESCRIPTION
**Pull request by escapedcat**
After looking at this please take a look at #1339 and merge that one first if possible.

### Describe the changes you have made in this PR

- For `Send - LNURLPay/confirmPayment` do not provide an `Origin`
- Hide `PublisherCard` if no `Origin` is set
- Handle `lnAddress` independent of `PublisherCard`
- Payment success notification only shows receiver if available
- Location in `TransactionList` will only be shown if available

### Link this PR to an issue

#770

### Type of change (Remove other not matching type)

- `fix`: Bug fix (non-breaking change which fixes an issue)

### Screenshots of the changes (If any)

- Via https://webln.twentyuno.net/invoice:
  image
- Via twitter:  
  image

### How has this been tested?

manually and added a basic unit-test for the success notification

### Checklist

- [x] My code follows the style guidelines of this project and performed a self-review of my own code
- [x] New and existing tests pass locally with my changes
- [x] I checked if I need to make corresponding changes to the documentation (and made those changes if needed)

### CODE REVIEW

File: src/app/screens/ConfirmPayment/index.tsx [hunk 1]
```diff
@@ -110,11 +114,14 @@ function ConfirmPayment(props: Props) {
       {!successMessage ? (
         <Container isScreenView maxWidth="sm">
           <div>
-            <PublisherCard
-              title={originRef.current.name}
-              image={originRef.current.icon}
-              url={originRef.current.host}
-            />
+            {!isOriginExtension && (
```
Comment (bumi) [inline src/app/screens/ConfirmPayment/index.tsx:117]: hmmm, 
ifs imo are always code-smells and add complexity.

**Review by bumi — COMMENTED** [review]
I am actually wondering if not dedicated UIs for the different cases is better?

File: src/app/screens/ConfirmPayment/index.tsx [hunk 2]
```diff
@@ -49,6 +50,9 @@ function ConfirmPayment(props: Props) {
   );
   const [fiatAmount, setFiatAmount] = useState("");
 
+  const isOriginExtension =
+    originRef?.current?.metaData?.url?.indexOf("chrome-extension") !== -1;
```
Comment (bumi) [inline src/app/screens/ConfirmPayment/index.tsx:54]: this looks a lot like a hack. 
and I don't expect this to work in other browsers.

if we don't have the origin we should also not pass it in. 
right now this is only done because the screens and actions need it. 
but ideally this dependency is not there - or what do you think? 

such a hack would need to be replicated in all the places.

File: src/app/screens/Send/index.tsx [hunk 3]
```diff
@@ -66,6 +67,8 @@ function Send() {
         }
 
         if (lnurlDetails.tag === "payRequest") {
+          originData.name = lnurlDetails.domain;
```
Comment (escapedcat) [inline src/app/screens/Send/index.tsx:70]: No sure if this is wanted. If we do not want this we could remove the origin completely here (and checking if this is used in other places i.e. `payment.location` etc)

File: src/app/screens/LNURLAuth/index.tsx [hunk 4]
```diff
@@ -23,13 +23,21 @@ function LNURLAuth() {
   const details = navState.args?.lnurlDetails as LNURLAuthServiceResponse;
   const origin = navState.origin;
 
+  if (!origin) {
```
Comment (reneaaron) [inline src/app/screens/LNURLAuth/index.tsx:26]: Is the origin also set if you paste the lnurl-auth string in the send screen?

File: src/app/screens/LNURLPay/index.tsx [hunk 5]
```diff
@@ -270,14 +273,21 @@ function LNURLPay() {
...
+            title={navState.origin.name}
+            description={navState.origin.description}
+            image={navState.origin.icon}
+            isSmall={false}
+          />
+        )}
         <dl className="shadow bg-white dark:bg-surface-02dp mt-4 pt-4 px-4 rounded-lg mb-6 overflow-hidden">
+          {getLnAddress() && (
+            <>
+              <Dt>LN Address</Dt>
```
Comment (reneaaron) [inline src/app/screens/LNURLPay/index.tsx:287]: If it's no problem space-wise I'd suggest to not abbreviate things and just write `Lightning Address`.

File: src/app/screens/LNURLAuth/index.tsx (hunk 4 above)
Comment (reneaaron) [reply src/app/screens/LNURLAuth/index.tsx:26]: This question applies to other methods (channel, withdraw) as well 👇

File: src/app/screens/ConfirmPayment/index.tsx [hunk 6]
```diff
@@ -66,7 +60,9 @@ function ConfirmPayment(props: Props) {
       const response = await utils.call(
         "sendPayment",
         { paymentRequest: paymentRequest },
-        { origin: navState.origin }
+        {
+          origin: { ...navState.origin, name: "" }, // name needs to be empty because we do not know the receiver
+        }
```
Comment (lisabaut) [inline src/app/screens/ConfirmPayment/index.tsx:65]: I am having trouble to understand this here.

`navState.origin` can be undefined if in Pop-Up.

But if in Prompt => `navState.origin` exists
And if so, then `navState.origin.name` exists too.
Why reset it then? 
Is the value incorrect for `name` ? 

If so, I would rephrase the comment too:

// always reset name to empty string, because ....

**Review by lisabaut — COMMENTED** [review]
Thanks!

For me the PR title is a bit misleading.

IMO what happens here is "Remove default origin data for LNURLPay and ConfirmPayment"

That the payment success notification has a fallback value in the end is mainly the result of this removal.

File: src/app/screens/LNURLPay/index.tsx [hunk 7]
```diff
@@ -207,18 +207,21 @@ function LNURLPay() {
     navigate(-1);
   }
 
-  function getRecipient() {
+  function getLnAddress() {
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/screens/LNURLPay/index.tsx]: Ah, didn't know that. What else can it be?
Reverted the name-change.

File: src/app/screens/Publisher.tsx [hunk 8]
```diff
@@ -68,13 +68,8 @@ function Publisher() {
...
               <p className="truncate">
-                <a
-                  target="_blank"
-                  title={payment.name}
-                  href={payment.location}
-                  rel="noreferrer"
-                >
-                  {payment.name}
+                <a target="_blank" href={payment.location} rel="noreferrer">
+                  {payment.name || payment.description || payment.host}
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/screens/Publisher.tsx:76]: Good point could tackle this in a follow up PR. I created a todo here and also will add a comment in the code:
https://github.com/getAlby/lightning-browser-extension/issues/1356

File: src/app/screens/Send/index.tsx [hunk 9]
```diff
@@ -89,10 +87,8 @@ function Send() {
       } else if (isPubKey(invoice)) {
         navigate(`/keysend?destination=${invoice}`);
       } else {
-        lightningPayReq.decode(invoice); // throws if invalid.
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/screens/Send/index.tsx:92]: Yup, thanks, didn't remove this on purpose.

File: src/types.ts [hunk 10]
```diff
@@ -110,7 +111,7 @@ export interface MessageDefault {
 }
 
 export type NavigationState = {
-  origin: OriginData;
+  origin?: OriginData;
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/types.ts]: Good point! Done

File: src/app/screens/ConfirmPayment/index.tsx (hunk 6 above)
Comment (escapedcat) [reply src/app/screens/ConfirmPayment/index.tsx:65]: You are right, this is confusing. I moved this.  

Context is that we only want `name` to be filled if we know the exact `receiver`. In `confirmPayment` we never know (correct me if I'm wrong @bumi @reneaaron ). So we can leave it out. 
A better place to do this is actually [persistPayments](https://github.com/getAlby/lightning-browser-extension/blob/fix/770_origin-receiver-issue/src/extension/background-script/events/persistPayments.js) because this is where it's saved. Did it there now.

File: src/app/router/Prompt/Prompt.tsx [hunk 11]
```diff
@@ -38,8 +38,11 @@ const createStateFromParams = (params: URLSearchParams): NavigationState => ({
 });
 
 const navigationState = createStateFromParams(params);
-
 function Prompt() {
+  if (!navigationState.origin) {
+    throw new Error("Origin is missing");
+  }
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/router/Prompt/Prompt.tsx:44]: This is only a guard for TS. Prompt always has an origin.  
Would it be better do adjust the `NavigationState` type here?  
Not sure how to handle this nicer.

File: src/app/screens/LNURLAuth/index.tsx [hunk 12]
```diff
@@ -23,13 +23,21 @@ function LNURLAuth() {
...
+    throw new Error("Origin is missing");
+  }
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/screens/LNURLAuth/index.tsx:28]: See above. This is a guard to handle the optional `origin`. LNURLAuth will always get an origin.
Maybe it's a better way to split `NavigationState` into one with and one without `origin`? What do you think?

File: src/app/screens/LNURLAuth/index.tsx [hunk 13]
```diff
@@ -23,13 +23,21 @@ function LNURLAuth() {
...
   const { t } = useTranslation("components", {
     keyPrefix: "confirmOrCancel",
   });
 
   const [successMessage, setSuccessMessage] = useState("");
 
   async function confirm() {
+    if (!origin) {
+      throw new Error("Origin is missing");
+    }
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/screens/LNURLAuth/index.tsx:39]: See above

File: src/app/screens/LNURLChannel/index.tsx [hunk 14]
```diff
@@ -17,8 +17,10 @@ import type { LNURLChannelServiceResponse } from "~/types";
 
 function LNURLChannel() {
   const navigate = useNavigate();
-
   const navState = useNavigationState();
+  if (!navState.origin) {
+    throw new Error("Origin is missing");
+  }
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/screens/LNURLChannel/index.tsx:23]: See above

File: src/app/screens/LNURLWithdraw/index.tsx [hunk 15]
```diff
@@ -20,6 +20,10 @@ function LNURLWithdraw() {
   const navigate = useNavigate();
   const navState = useNavigationState();
 
+  if (!navState.origin) {
+    throw new Error("Origin is missing");
+  }
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/screens/LNURLWithdraw/index.tsx:25]: See above

File: src/app/router/Prompt/Prompt.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (lisabaut) [reply src/app/router/Prompt/Prompt.tsx:44]: If this is only for TS, then I would remove all error throws. 

Here in `Prompt` you know that it is defined.

So in all occasions in this file I would do an assertion:
```TS
            <Route
              path="webln/enable"
              element={<Enable origin={navigationState.origin as OriginData} />} // assertion !
            />
```

In this file this assertion will be obsolete later anyways when all components are refactored to not use searchparams no more.

File: src/app/screens/LNURLAuth/index.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (lisabaut) [reply src/app/screens/LNURLAuth/index.tsx:39]: In this PR `origin` is always defined for LNURLAuth (I know you change it later in #1339)

No need to do a check here.

Rather do this:
`const origin = navState.origin as OriginData; // always defined for LNURLAuth via Prompt or Pop-Up`

File: src/app/screens/LNURLAuth/index.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (lisabaut) [reply src/app/screens/LNURLAuth/index.tsx:28]: I would do a type refactoring later when all origins are removed where they should not be. To get an overview.
Until then all components which can be rendered with or without origin should make that clear.

Either with a type assertion where you know that origin always exists:
`const origin = navState.origin as OriginData; // always defined via Prompt or Pop-Up`

or when the component should handle both origin defined or undefined with all checks
`if (navState.origin) { ....` , the way you already did it.

File: src/app/screens/LNURLChannel/index.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (lisabaut) [reply src/app/screens/LNURLChannel/index.tsx:23]: Thanks. It is very unusal to throw in the middle of a component. Especially when this is not handled nowhere (show Toast, show Error component ....)
And if this is only for TS then I would make this more obvious. See comments above.

File: src/app/router/Prompt/Prompt.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/router/Prompt/Prompt.tsx:44]: Got it, I'm so focused on avoiding `as` that I don't even consider it anymore. Will use this.

File: src/app/screens/LNURLAuth/index.tsx [hunk 16]
```diff
@@ -14,14 +14,14 @@ import { USER_REJECTED_ERROR } from "~/common/constants";
...
-import type { LNURLAuthServiceResponse } from "~/types";
+import type { LNURLAuthServiceResponse, OriginData } from "~/types";
 
 function LNURLAuth() {
   const navigate = useNavigate();
 
   const navState = useNavigationState();
   const details = navState.args?.lnurlDetails as LNURLAuthServiceResponse;
-  const origin = navState.origin;
+  const origin = navState.origin as OriginData;
```
Comment (lisabaut) [inline src/app/screens/LNURLAuth/index.tsx:24]: I would add a small comment here `// origin is always defined for LNURLAuth`
For our future selves.

File: src/app/screens/ConfirmPayment/index.tsx (hunk 6 above)
Comment (bumi) [reply src/app/screens/ConfirmPayment/index.tsx:65]: I think that's not correct. 
the origin data comes from the website on which the user interacts. So for actions that are triggered from interacting on the website (e.g. webln.sendPayment()) we have metadata. 
if a user manually enters an invoice then we do not have any metadata.  - and we get the receiver name/logo from that website metadata.

why do we never know the receiver in the confirmPayment case? - I think we do in the case described above.

Comment (escapedcat) [reply src/app/screens/ConfirmPayment/index.tsx:65]: Depends on the definition of "receiver". Let's talk.

Comment (bumi) [reply src/app/screens/ConfirmPayment/index.tsx:65]: hmm, not sure what definition we need? did you look at the code?

**Comment by bumi** [comment]
@dvoroneca is this payment confirmation screen ok for now? 
image

File: src/extension/background-script/events/notifications.ts [hunk 17]
```diff
@@ -0,0 +1,79 @@
+import utils from "~/common/lib/utils";
+import type { PaymentNotificationData } from "~/types";
+
+const paymentSuccessNotification = (
+  message: "ln.sendPayment.success",
```
Comment (bumi) [inline src/extension/background-script/events/notifications.ts:5]: this can also be a different string, can't it?

File: src/extension/background-script/events/persistPayments.js [hunk 18]
```diff
@@ -1,16 +1,16 @@
 import db from "../db";
 
 const persistSuccessfullPayment = async (message, data) => {
-  const recipientName = data.origin.name;
-  const host = data.origin.host;
+  const host = data?.origin?.host || "";
```
Comment (bumi) [inline src/extension/background-script/events/persistPayments.js:4]: do we want the host to be an empty string? is this a good practice or should it rather be something null? 
I assume the empty string is to make TS happy but wouldn't a null value be more useful in a DB?

File: src/extension/background-script/events/persistPayments.js [hunk 19]
```diff
@@ -1,16 +1,16 @@
...
   const route = paymentResponse.data.route;
   const { total_amt, total_fees } = route;
 
   await db.payments.add({
-    host: host,
-    location: data.origin.location,
-    name: recipientName,
+    host,
+    location,
+    name: "", // we never know the "receiver" when using `confirmPayment`. To avoid setting a website as "receiver" we omit this.
```
Comment (bumi) [inline src/extension/background-script/events/persistPayments.js:13]: I don't understand this. 
and this is also different to how it currently is. we stored the `recipientName` which is `data.origin.name`;

**Comment by dvoroneca** [comment]
Yes, it looks a bit empty. Here is a confirmation card design proposed in the 1-click pay flow. 

Successful transaction V1+
Successful transaction V1

**Review by lisabaut — APPROVED** [review]
Thanks!

Again, in my opinion this PR is actually doing this:
"refactor: remove default origin for LNURL`

Because you remove `getOriginData` completely from the `Send` component and its navigation to all LNURL component (Auth, Withdraw, Channel, Pay) in this PR.

Please see my comments below. 
There is still `origin` defined in LNURLAuth which can be optional when coming via `Send`.

File: src/app/screens/LNURLAuth/index.tsx [hunk 20]
```diff
@@ -14,14 +14,15 @@ import { USER_REJECTED_ERROR } from "~/common/constants";
...
+import type { LNURLAuthServiceResponse, OriginData } from "~/types";
 
 function LNURLAuth() {
   const navigate = useNavigate();
 
   const navState = useNavigationState();
+
   const details = navState.args?.lnurlDetails as LNURLAuthServiceResponse;
-  const origin = navState.origin;
+  const origin = navState.origin as OriginData; // this action will always have an `origin` set, just the type is optional to support usage via PopUp
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/screens/LNURLAuth/index.tsx]: Yup, thanks!

File: src/types.ts [hunk 21]
```diff
@@ -276,7 +288,7 @@ export interface MessageConnectPeer extends MessageDefault {
 
 export interface MessageLnurlAuth {
   args: {
-    origin: OriginData;
+    origin?: OriginData; // only set if troggered via Prompt
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/types.ts]: 😊

File: src/app/screens/LNURLPay/index.tsx [hunk 22]
```diff
@@ -271,12 +274,11 @@ function LNURLPay() {
     return (
       <Container maxWidth="sm">
         <PublisherCard
-          title={navState.origin.name}
-          description={navState.origin.description}
-          lnAddress={getRecipient()}
-          image={navState.origin.icon}
-          isSmall={false}
+          title={getRecipient()}
+          description={navState?.origin?.description}
```
[the remark this answers is not in the document]
Comment (escapedcat) [reply src/app/screens/LNURLPay/index.tsx]: Yup, thanks!
